gpt_explainer/gpt_slide_expander.py

```diff
@@ -0,0 +1,272 @@
+import openai
+
+
+class GptSlideExpander:
+    """
+    A class that utilizes OpenAI's GPT models to generate expanded explanations for presentation slides.
+
+    Attributes:
+    expanded_slide_explanations (Dict[int, str]): A dictionary that maps slide indices to their generated explanations.
+    The keys represent the slide index, and the values are the corresponding expanded explanations for each slide.
+    This dictionary is populated when `generate_explanation_for_presentation` is called.
+    """
+
+    def __init__(self):
+        GptSlideExpander.configure_api_key()
+        self.expanded_slide_explanations = {}
+        self.processed_files_lock = threading.Lock()
+        self.uploads_folder = ""
+        self.output_folder = ""
+
+    def set_output_folder(self, output_path: str) -> None:
+        self.output_folder = output_path
+
+    def set_uploads_folder(self, uploads_path: str) -> None:
+        self.uploads_folder = uploads_path
+
+    @staticmethod
+    def configure_api_key():
+        """Configure OpenAI API key."""
+        openai.api_key = os.getenv("OPENAI_API_KEY")
+
+    @staticmethod
+    def generate_presentation_main_topic(presentation_text: str, max_retry: int) -> str:
+        """
+        Generates the main topic of a presentation based on the presentation's text.
+        Args:
+            presentation_text (str): The text of the presentation.
+            max_retry (int): The maximum number of retries for generating the main topic.
+        Returns:
+            str: The generated main topic of the presentation.
+        Raises:
+            RuntimeError: If the query fails after reaching the maximum number of retries.
+            ValueError: If the presentation text is empty or invalid.
+        """
+        if presentation_text:
+            retry_count = 0
+            wait_time = 4  # Initial wait time in seconds
+
+            while retry_count < max_retry:
+                try:
+                    gpt_response = openai.ChatCompletion.create(
+                        model="gpt-3.5-turbo",
+                        messages=[{"role": "system",
+                                   "content": "You are a chatbot that extracts the main topic of a presentation"
+                                              " based on the presentation's text."},
+                                  {"role": "user",
+                                   "content": f"Provide the main topic of this presentation based on the presentation's"
+                                              f" following text: {presentation_text} , no longer than one word."}]
+
+                    )
+
+                    return gpt_response['choices'][0]['message']['content']
+
+                except openai.error as e:
+                    pass  # error will be documented in log files in future
+
+                # Increase wait time and retry
+                retry_count += 1
+                time.sleep(wait_time)
+                wait_time += 1
```

Comment:
Assign constants to descriptive variables to enhance readability and simplify future changes ^__^ . Directly using them can lead to confusion and maintenance challenges.